include/rmm/mr/device/logging_resource_adaptor.hpp

```diff
@@ -0,0 +1,204 @@
+    logger_->info(csv_header);
+    logger_->set_pattern("%H:%M:%S:%f,%v");
+  }
+
+  /**
+   * @brief Return pointer to the upstream resource.
+   *
+   * @return Upstream* Pointer to the upstream resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @brief Checks whether the upstream resource supports streams.
+   *
+   * @return true The upstream resource supports streams
+   * @return false The upstream resource does not support streams.
+   */
+  bool supports_streams() const noexcept override {
+    return upstream_->supports_streams();
+  }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource and logs the allocation.
+   *
+   * If the upstream allocation is successful logs the
+   * following CSV formatted line to the file specified at construction:
+   * ```
+   * *TIMESTAMP*,"allocate",*bytes*,*stream*
+   * ```
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    auto const p = upstream_->allocate(bytes, stream);
+    std::string msg{"allocate,"};
+    std::stringstream ss;
+    ss << p;
+    msg += ss.str();
+    msg += ",";
+    msg += std::to_string(bytes);
+    msg += ",";
+    msg += std::to_string(reinterpret_cast<uintptr_t>(stream));
+    logger_->info(msg);
+    return p;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `p` and log the
+   * deallocation.
+   *
+   * Every invocation of `logging_resource_adaptor::do_deallocate` will write
+   * the following CSV formatted line to the file specified at construction:
+   * ```
+   * *TIMESTAMP*,"free",*bytes*,*stream*
+   * ```
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override {
+    std::string msg{"free,"};
+    std::stringstream ss;
+    ss << p;
+    msg += ss.str();
+    msg += ",";
+    msg += std::to_string(bytes);
+    msg += ",";
+    msg += std::to_string(reinterpret_cast<uintptr_t>(stream));
+    logger_->info(msg);
+    upstream_->deallocate(p, bytes, stream);
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept {
+    return upstream_->is_equal(other);
```

Comment:
I realized while working on `thread_safe_resource_adaptor` that this approach causes the `SelfEquality` test to fail, because it calls `mr->is_equal(mr)`, which would call `mr->upstream()->is_equal(mr)`

Reply:
BTW, this seems to work (passes the test):

```
/**
   * @brief Compare the upstream resource to another.
   *
   * @throws Nothing.
   *
   * @param other The other resource to compare to
   * @return true If the two resources are equivalent
   * @return false If the two resources are not equivalent
   */
  bool do_is_equal(device_memory_resource const& other) const noexcept {
    if (this == &other) return true;
    else {
      thread_safe_resource_adaptor<Upstream> const* thread_safe_other = 
        dynamic_cast<thread_safe_resource_adaptor<Upstream> const*>(&other);
      if (thread_safe_other == nullptr) return false;
      else return upstream_->is_equal(*thread_safe_other->get_upstream());
    }
  }
```

Reply:
Done.

Reply:
Indentation is ugly, but OK.